README.md

# PalCreatoR <img src = "man/figures/Hex Sticker.png" align = "right" height = "200" />

<!-- badges: start -->

<!-- badges: end -->

The PalCreatoR package is designed to create an R palette using colors
from an image. It streamlines the process of pulling out a number of
representative colors from the given image and returning the
corresponding hexadecimal codes. It also offers the option to get
colorblind-friendly colors. Just pick an image of your favorite and pass
it into the
[function](https://genchanghsu.github.io/PalCreatoR/reference/create_pal.html):
a customized palette is readily available for you\! Use it for
presentation, publication, or simply for fun\!

## Background

The idea of creating color palettes from an image is not something new
in the R community. Several packages (e.g.,
[imgpalr](https://cran.rstudio.com/web/packages/imgpalr/index.html),
[paletteR](https://github.com/AndreaCirilloAC/paletter),
[PNWColors](https://github.com/jakelawlor/PNWColors),
[RImagePalette](https://cran.r-project.org/web/packages/RImagePalette/index.html))
and functions (e.g.,
[palettebuildr](https://gist.github.com/jonesor/1818babb03783dc41a1a))
have been developed for this purpose, with a wide range of arguments
available. Inspired by these works, the PalCreatoR package aims to
provide an easy-to-use tool that does the same job while keeping the
arguments as simple as possible. It incorporates two features that have
so far not yet been implemented in the current functions of its kind:

</p>

<p>

1)  Allowing user to get colorblind-friendly colors converted from the
    original colors in the palette.

</p>

<p>

2)  Using multivariate Gaussian mixture modeling (GMM), besides the
    most-used kmeans algorithm, to extract the representative colors
    from the image.

## Installation

The latest version of PalCreatoR can be installed from GitHub:

``` r
# install.packages("devtools")
devtools::install_github("GenChangHSU/PalCreatoR")
```

## Examples

Here are some example palettes created with the function
[`create_pal`](https://genchanghsu.github.io/PalCreatoR/reference/create_pal.html).
See [Quick
Demonstration](https://genchanghsu.github.io/PalCreatoR/articles/Quick_Demonstration.html)
for more details on its usage.

<br>

<img src = "inst/Mountain.JPG" align = "left" height = "300" />
<img src="man/figures/README-palette example 1-1.png" width="30%" />

<br> <br>

<img src = "inst/Cicada.JPG" align = "left" height = "300" />
<img src="man/figures/README-palette example 2-1.png" width="30%" />

<br> <br>

<img src = "inst/Alpine Flower.JPG" align = "left" height = "300" />
<img src="man/figures/README-palette example 3-1.png" width="30%" />

<br>

## Contact

Like it? Feel free to reach us at <genchanghsu@gmail.com> and
<willemou@gmail.com>. Any suggestions or cool ideas are welcome\!


NEWS.md
# PalCreatoR 1.0.0 (2021-04-01)

* Created a new function [modify_pal](https://genchanghsu.github.io/PalCreatoR/reference/modify_pal.html) for the package

* Updated the [Quick Demonstration](https://genchanghsu.github.io/PalCreatoR/articles/Quick_Demonstration.html) and added some example usages of [modify_pal](https://genchanghsu.github.io/PalCreatoR/reference/modify_pal.html) 

# PalCreatoR 0.2.1 (2021-03-01)

* Added "Example Images" section to the [Functions](https://genchanghsu.github.io/PalCreatoR/reference/index.html) page

# PalCreatoR 0.2.0 (2021-01-05)

* Completed the [Quick Demonstration](https://genchanghsu.github.io/PalCreatoR/articles/Quick_Demonstration.html) of the package

# PalCreatoR 0.1.0 (2021-01-03)

* Added a `NEWS.md` file to track changes to the package


# PalCreatoR 0.0.1 (2021-01-01)

* Initial release on GitHub






R/modify_pal.R
#' @title Modify alpha transparency of the colors in the palette
#'
#' @description \code{modify_pal} modifies alpha transparency of the colors in the palette.
#'
#' @param pal a vector of hexadecimal color codes (not necessarily returned from \code{\link[PalCreatoR:create_pal]{create_pal}).
#' @param alpha a single number or a vector of numbers between 0 and 1. These values define the degree
#'     of transparency of the colors in the palette. If \code{alpha} is a single number, the transparency of
#'     all the colors in the palette will be set to that value; if \code{alpha} is a vector of numbers, the
#'     transparency of the colors in the palette will be set to the corresponding alpha values.
#'     Also note that if the vector lengths of \code{pal} and \code{alpha} differ, extra elements in the longer
#'     vector will be omitted to match the length of the shorter one. See 'Details' section for more information
#'     on the concept of alpha transparency.
#' @param show.pal logical. Whether to display the modified palette or not. Default to \code{TRUE}.
#' @param title a character string giving the title of the displayed palette.
#' @param ... additional arguments passed to \code{\link[ggplot2:theme]{ggplot2::theme}}.
#'
#' @details An alpha value defines the "transparency", or "opacity" of the color. A value of 0 means completely
#'     transparent (i.e., the background will completely “show through”); a value of 1 means completely opaque
#'     (i.e., none of the background will “show through”). In short, the lower the alpha value is, the lower "amount"
#'     of the color will be.
#'
#' @return A vector of hexadecimal color codes with two additional digits defining the degree of transparency.
#'
#' @importFrom magrittr "%>%"
#'
#' @examples \dontrun{
#' library(PalCreatoR)
#' image_path <- system.file("Mountain.JPG", package = "PalCreatoR")
#'
#' My_pal <- create_pal(image = image_path,
#'                      n = 5,
#'                      resize = 0.1,
#'                      method = "kmeans",
#'                      colorblind = FALSE,
#'                      sort = "value",
#'                      show.pal = TRUE,
#'                      title = "My Palette")
#'
#' My_new_pal <- modify_pal(pal = My_pal,
#'                          alpha = c(0.2, 0.4, 0.6, 0.8, 1.0),
#'                          show.pal = TRUE,
#'                          title = "My New Palette")
#' print(My_new_pal)}
modify_pal <- function(pal,
                       alpha,
                       show.pal = TRUE,
                       title = "",
                       ...) {
  # Error messages -------------------------------------------------------------------------

  # 1. pal argument
  color_check <- sapply(pal, function(X){
    grepl(pattern = "^#[0-9A-Fa-f]{6}$", x = X)
  })

  if (any(color_check == F)) {
    stop('One or more incorrect hex color codes passed in the "pal" argument!')
  }

  # 2. alpha argument
  if (any(!alpha <= 1) || any(!alpha >= 0)) {
    stop('One or more incorrect values passed in the "alpha" argument!')
  }

  # 3. show.pal argument
  if (is.logical(show.pal) == F) {
    stop('Argument passed to "show.pal" is not logical!')
  }


  # Function body -----------------------------------------------------------

  # 1. Check the lengths of the pal and alpha vectors
  if (length(pal) != length(alpha) &&
      length(pal) != 1 && length(alpha) != 1) {

    l_pal <- length(pal)
    l_alpha <- length(alpha)
    n_shorter <- min(l_pal, l_alpha)
    df <- data.frame(hex = pal[1:n_shorter], alpha = alpha[1:n_shorter])

    warning(
      'The lengths of "pal"" and "alpha" differ; extra elements in the longer vector
  are omitted to match the length of the shorter one!'
    )
  } else {
    df <- data.frame(hex = pal, alpha = alpha)
  }

  # 2. Get the hex codes with the additional two alpha digits
  hex_codes <- purrr::map2(
    .x = df$hex,
    .y = df$alpha,
    .f = function(x, y) {
      rgb_val <- col2rgb(x, alpha = F) %>% as.vector()
      hex_code <-
        rgb(
          r = rgb_val[1],
          g = rgb_val[2],
          b = rgb_val[3],
          alpha = y * 255,
          maxColorValue = 255
        )
      return(hex_code)
    }
  ) %>% unlist()

  # 3. Visualize the palette
  n <- length(hex_codes)

  if (show.pal == T) {
    if (n <= 10) {
      Pal_df <- hex_codes %>%
        data.frame(Hex_code = .) %>%
        dplyr::mutate(x = rep(1, n),
                      y = 10:(10 - n + 1))
    }

    if (n > 10 & n %% 10 != 0) {
      q <- n %/% 10
      m <- n %% 10

      Pal_df <- hex_codes %>%
        data.frame(Hex_code = .) %>%
        dplyr::mutate(x = c(rep(1:q, each = 10), rep(q + 1, m)),
                      y = c(rep(10:1, q), 10:(10 - m + 1)))
    }

    if (n > 10 & n %% 10 == 0) {
      q <- n %/% 10

      Pal_df <- hex_codes %>%
        data.frame(Hex_code = .) %>%
        dplyr::mutate(x = c(rep(1:q, each = 10)),
                      y = c(rep(10:1, q)))
    }

    p <- ggplot2::ggplot(Pal_df, ggplot2::aes(x = x, y = y)) +
      ggplot2::geom_tile(ggplot2::aes(fill = Hex_code)) +
      ggplot2::geom_label(ggplot2::aes(label = Hex_code),
                          fill = "grey",
                          size = 4) +
      ggplot2::scale_fill_identity() +
      ggplot2::theme_void() +
      ggplot2::labs(title = title) +
      ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5, size = 18)) +
      ggplot2::theme(...)

    print(p)
  }

  # 9. Return the palette vector
  return(hex_codes)
}













R/example_images.R
#' @title Example images included in this package
#'
#' @description Three example images (Mountain.JPG, Cicada.JPG, Alpine Flower.JPG)
#'      are available in this package (photo credit: Gen-Chang Hsu).
#'      See 'Examples' for how to load the images.
#'
#' @format .JPG file.
#'
#' @examples
#' library(PalCreatoR)
#' library(magick)
#'
#' image_path_Mountain <- system.file("Mountain.JPG", package = "PalCreatoR")
#' Mountain <- image_read(image_path_Mountain)
#' plot(Mountain)
#'
#' image_path_Cicada <- system.file("Cicada.JPG", package = "PalCreatoR")
#' Cicada <- image_read(image_path_Cicada)
#' plot(Cicada)
#'
#' image_path_Alpine_flower <- system.file("Alpine Flower.JPG", package = "PalCreatoR")
#' Alpine_flower <- image_read(image_path_Alpine_flower)
#' plot(Alpine_flower)
#'
"example_images"


Hex Sticker/Hex Sticker.R
########## Hexagon Sticker for the PalCreatoR Package ###########
########## Author: Gen-Chang Hsu ##########

library(tidyverse)
library(hexSticker)
library(magick)

### Sticker subplot
spiral_df <- data.frame(x = seq(0, 13, length.out = 5000)) %>%
  mutate(y = exp(0.30635 * x)) %>%
  mutate(
    x = rev(x),
    y = rev(y),
    polar_x = y * cos(x),
    polar_y = y * sin(x),
    x_start = rep(0, 5000),
    y_start = rep(0, 5000)
  )

spiral_p <- ggplot(spiral_df) +
  geom_segment(aes(x = x_start, y = y_start, xend = polar_x, yend = polar_y),
    color = rev(rainbow(5000)),
    size = 1
  ) +
  geom_point(aes(x = polar_x, y = polar_y),
    color = "black",
    size = 0.1
  ) +
  coord_fixed(ratio = 1) +
  theme_void()

### Create the hex sticker
sticker(

  # Sticker image
  subplot = spiral_p,
  s_x = 1,
  s_y = 1.1,
  s_width = 1.1,
  s_height = 2.1,

  # Package name
  package = "PalCreatoR",
  p_x = 1,
  p_y = 0.5,
  p_color = "black",
  p_family = "Aller_Rg",
  p_fontface = "bold",
  p_size = 36,

  # Hexagon background
  h_size = 1.8,
  h_fill = "#9BCD9B",
  h_color = "black",

  # Spotlight
  spotlight = F,
  l_x = 1,
  l_y = 0.5,
  l_width = 3,
  l_height = 3,
  l_alpha = 0.4,

  # URL
  url = "",
  u_x = 1,
  u_y = 0.08,
  u_color = "black",
  u_family = "Aller_Rg",
  u_size = 1.5,
  u_angle = 30,

  # Sticker border
  white_around_sticker = F,

  # Save the file
  filename = "man/figures/Hex Sticker.png",
  asp = 1,
  dpi = 600
)


R/create_pal.R
#' @title Create a palette using colors extracted from an image
#'
#' @description \code{create_pal} creates a palette using
#'     colors extracted from an image.
#'
#' @param image a file path/URL to the raster image (JPG, JPEG, PNG, TIFF) from which
#'     the colors are to be extracted.
#' @param n a positive integer. The number of colors to be returned in the palette.
#' @param resize a number between 0 and 1. This indicates the fraction by which
#'     the width and height (in pixels) of the original image is resized through
#'     \code{\link[magick:image_resize]{magick::image_resize}}. Default to 0.1.
#' @param method the classification method for grouping the image pixels based
#'     on the RGB values. Options are \code{"kmeans"} and \code{"Gaussian_mix"}.
#'     Default to \code{"kmeans"}. See 'Details' section for more information.
#' @param colorblind logical. Whether to render the palette colorblind-friendly.
#'     Default to \code{FALSE}. See 'Details' section for more information.
#' @param sort a character indicating how the colors in the palette should be sorted.
#'     Available options are \code{"none"}, \code{"hue"}, \code{"saturation"},
#'     and \code{"value"}. Default to \code{"none"}(unsorted). See 'Details' section
#'     for more information.
#' @param show.pal logical. Whether to display the palette or not. Default to \code{TRUE}.
#' @param title a character string giving the title of the displayed palette.
#' @param ... additional arguments passed to \code{\link[ggplot2:theme]{ggplot2::theme}}.
#'
#' @details Two clustering methods are available. For \code{method = "kmeans"},
#'     image pixels are partitioned into clusters using \code{\link[stats:kmeans]{kmeans}},
#'     and the RGB values of the cluster centers are converted into the corresponding
#'     hexadecimal color codes. For \code{method = "Gaussian_mix"}, pixel components
#'     are identified via multivariate Gaussian mixture modeling using \code{\link[ClusterR:GMM]{ClusterR::GMM}},
#'     and the RGB values of the component centroids are converted into the
#'     corresponding hexadecimal color codes.
#'
#'     If \code{"colorblind = TRUE"}, the original colors returned in the palette
#'     are converted into their colorblind-friendly counterparts using
#'     \code{\link[colorBlindness:replacePlotColor]{colorBlindness::replacePlotColor}}.
#'
#'     The colors returned in the palette can be ordered in the HSV color space.
#'     If \code{sort = "hue"}, the colors are sorted by hue in an ascending order.
#'     If \code{sort = "saturation"}, the colors are sorted by saturation in a descending order.
#'     If \code{sort = "value"}, the colors are sorted by value in a descending order.
#'
#' @return A vector of hexadecimal color codes.
#'
#' @importFrom magrittr "%>%"
#'
#' @examples \dontrun{
#' library(PalCreatoR)
#' image_path <- system.file("Mountain.JPG", package = "PalCreatoR")
#'
#' My_pal <- create_pal(image = image_path,
#'                      n = 5,
#'                      resize = 0.1,
#'                      method = "kmeans",
#'                      colorblind = FALSE,
#'                      sort = "value",
#'                      show.pal = TRUE,
#'                      title = "My Palette")
#'
#' print(My_pal)}
create_pal <- function(image,
                       n,
                       resize = 0.1,
                       method = "kmeans",
                       colorblind = FALSE,
                       sort = "none",
                       show.pal = TRUE,
                       title = "",
                       ...) {

  # Error messages -------------------------------------------------------------------------

  # 1. n argument
  is.wholenumber <- function(x, tol = .Machine$double.eps^0.5) {
    abs(x - round(x)) < tol
  }

  if (is.wholenumber(n) == F || n <= 0) {
    stop("Incorrect n value. Use positive integer only!")
  }

  # 2. resize argument
  if (!resize <= 1 || !resize >= 0) {
    stop("Incorrect resize value!")
  }

  # 3. method argument
  if (!method %in% c("kmeans", "Gaussian_mix")) {
    stop("Incorrect clustering method!")
  }

  # 4. colorblind argument
  if (is.logical(colorblind) == F) {
    stop('Argument passed to "colorblind" is not logical!')
  }

  # 5. sort argument
  if (!sort %in% c("none", "hue", "saturation", "value")) {
    stop("Unknown sorting method!")
  }

  # 6. title argument
  if (is.character(title) == F) {
    stop("Incorrect title for the palette!")
  }

  # 7. show.pal argument
  if (is.logical(show.pal) == F) {
    stop('Argument passed to "show.pal" is not logical!')
  }


  # Function body -----------------------------------------------------------

  # 1. Read the image
  img <- magick::image_read(image)

  # 2. Get the width and height of the image
  img_width <- magick::image_info(img)["width"]
  img_height <- magick::image_info(img)["height"]


  # 3. Resize the image to reduce the computation load
  img_resize <- magick::image_resize(
    img,
    magick::geometry_size_pixels(
      width = img_width * resize,
      height = img_height * resize,
      preserve_aspect = TRUE
    )
  )


  # 4. Convert the image object into a dataframe of RGB values
  RGB_raw <- imager::magick2cimg(img_resize) %>%
    as.data.frame(wide = "c") %>%
    dplyr::rename(Red = c.1, Green = c.2, Blue = c.3)


  # 5. Cluster analysis on the RGB values
  # 5.1 kmeans clustering
  if (method == "kmeans") {
    set.seed(123)
    kmeans_out <- kmeans(RGB_raw[, c("Red", "Green", "Blue")], centers = n)
    if (kmeans_out$ifault == 4) {
      kmeans_out <-  kmeans(RGB_raw[, c("Red", "Green", "Blue")],
                            centers = n,
                            algorithm = "MacQueen")
      }

    RGB_values <- kmeans_out$centers %>%
      as.data.frame() %>%
      dplyr::mutate(
        Cluster = 1:n,
        Hex_code = grDevices::rgb(
          red = Red,
          green = Green,
          blue = Blue
        )
      )
  }

  # 5.2 Gaussian mixture modeling
  if (method == "Gaussian_mix") {
    GMM_out <- ClusterR::GMM(RGB_raw[, c("Red", "Green", "Blue")],
      gaussian_comps = n,
      dist_mode = "eucl_dist",
      seed_mode = "random_subset",
      km_iter = 10,
      em_iter = 10,
      verbose = F
    )
    RGB_values <- GMM_out$centroids %>%
      as.data.frame() %>%
      dplyr::rename(Red = V1, Green = V2, Blue = V3) %>%
      dplyr::mutate(
        Cluster = 1:n,
        Hex_code = grDevices::rgb(
          red = Red,
          green = Green,
          blue = Blue
        )
      )
  }


  # 6. Make the original colors colorblind-friendly
  if (colorblind == T) {
    Pal_vector <- colorBlindness::displayColors(RGB_values$Hex_code) %>%
      colorBlindness::replacePlotColor() %>%
      .[["grobs"]] %>%
      .[[6]] %>%
      .[["children"]] %>%
      .[[3]] %>%
      .[["gp"]] %>%
      .[["fill"]] %>%
      stringr::str_sub(., start = 1, end = 7)
  } else {
    Pal_vector <- RGB_values$Hex_code
  }


  # 7. Sort the colors based on their HSV
  if (sort == "none") {
    Pal_vector <- Pal_vector
  }

  if (sort == "hue") {
    Pal_vector <- grDevices::col2rgb(Pal_vector) %>%
      grDevices::rgb2hsv() %>%
      t() %>%
      as.data.frame() %>%
      dplyr::arrange(h) %>%
      dplyr::transmute(Pal = hsv(h, s, v)) %>%
      dplyr::pull(Pal)
  }

  if (sort == "saturation") {
    Pal_vector <- grDevices::col2rgb(Pal_vector) %>%
      grDevices::rgb2hsv() %>%
      t() %>%
      as.data.frame() %>%
      dplyr::arrange(desc(s)) %>%
      dplyr::transmute(Pal = hsv(h, s, v)) %>%
      dplyr::pull(Pal)
  }

  if (sort == "value") {
    Pal_vector <- grDevices::col2rgb(Pal_vector) %>%
      grDevices::rgb2hsv() %>%
      t() %>%
      as.data.frame() %>%
      dplyr::arrange(desc(v)) %>%
      dplyr::transmute(Pal = hsv(h, s, v)) %>%
      dplyr::pull(Pal)
  }


  # 8. Visualize the palette
  if (show.pal == T) {

  if (n <= 10) {
    Pal_df <- Pal_vector %>%
      data.frame(Hex_code = .) %>%
      dplyr::mutate(
        x = rep(1, n),
        y = 10:(10 - n + 1)
      )
  }

  if (n > 10 & n %% 10 != 0) {
    q <- n %/% 10
    m <- n %% 10

    Pal_df <- Pal_vector %>%
      data.frame(Hex_code = .) %>%
      dplyr::mutate(
        x = c(rep(1:q, each = 10), rep(q + 1, m)),
        y = c(rep(10:1, q), 10:(10 - m + 1))
      )
  }

  if (n > 10 & n %% 10 == 0) {
    q <- n %/% 10

    Pal_df <- Pal_vector %>%
      data.frame(Hex_code = .) %>%
      dplyr::mutate(
        x = c(rep(1:q, each = 10)),
        y = c(rep(10:1, q))
      )
  }

  p <- ggplot2::ggplot(Pal_df, ggplot2::aes(x = x, y = y)) +
    ggplot2::geom_tile(ggplot2::aes(fill = Hex_code)) +
    ggplot2::geom_label(ggplot2::aes(label = Hex_code), fill = "grey", size = 5) +
    ggplot2::scale_fill_identity() +
    ggplot2::theme_void() +
    ggplot2::labs(title = title) +
    ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5, size = 18)) +
    ggplot2::theme(...)

    print(p)
  }

  # 9. Return the palette vector
  return(Pal_vector)
}


